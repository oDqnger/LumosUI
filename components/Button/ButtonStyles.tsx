export const buttonColors = {
    "default": {
        normal: "bg-zinc-500 ",
        clicked: "bg-white",
        disable: "bg-zinc-700 ",
        hover: "hover:bg-zinc-600"
    },
    "primary": {
        normal: "bg-sky-500 ",
        clicked: "bg-white",
        disable: "bg-sky-700 ",
        hover: "hover:bg-sky-600"
    },
    "secondary": {
        normal: "bg-violet-600 ",
        clicked: "bg-white",
        disable: "bg-violet-800 ",
        hover: "hover:bg-violet-700"
    },
    "success": {
        normal: "bg-green-500 ",
        clicked: "bg-white",
        disable: "bg-green-700 ",
        hover: "hover:bg-green-600"
    },
    "warning": {
        normal: "bg-amber-600 ",
        clicked: "bg-white",
        disable: "bg-amber-800 ",
        hover: "hover:bg-amber-700"
    },
    "danger": {
        normal: "bg-red-600 ",
        clicked: "bg-white",
        disable: "bg-red-800 ",
        hover: "hover:bg-red-700"
    }
}

export const radiusStyles = {
    "none": "rounded-none ",
    "sm": "rounded-sm ",
    "md": "rounded ",
    "lg": "rounded-lg ",
    "full": "rounded-full "
}

export const buttonSizes = {
    "sm": "pr-5 pl-5 pt-1 pb-1 text-sm ",
    "md": "",
    "lg": "pr-8 pl-8 pt-3 pb-3 text-lg ",
}

export const baseStyles = "border pr-7 pl-7 pt-2 pb-2 transition ease-in-out "
