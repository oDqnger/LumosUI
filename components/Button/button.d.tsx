import { ButtonHTMLAttributes } from "react";

type ColorTypes = "default" | "primary" | "secondary" | "success" | "warning" | "danger"
type RadiusTypes = "none" | "sm" | "md" | "lg" | "full"
type SizeTypes = "sm" | "md" | "lg"

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    isDisabled?: boolean,
    overrideDefaultOnClick?: Function
}

interface DefaultProps {
    children?: any,
    className?: string,
}

interface StyledProps {
    color?: ColorTypes,
    radius?: RadiusTypes,
    size?: SizeTypes,
    customStyles?: string,
}

export type AllProps = ButtonProps & DefaultProps & StyledProps;
