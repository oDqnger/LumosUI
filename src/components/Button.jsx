import PropTypes from 'prop-types'

// TODO - add radius option
// TODO - add colors (with the ability to use your own custom colors.)
// TODO - add variants

export default function Button(props) {

    const {
        radius="lg",
        onClick: handleClick,
        size="md", 
        isDisabled=false, 
        children: text, 
        ...defaultProps
        } = props;

    const radiusStyles = {
        "none": "rounded-none",
        "sm": "rounded-sm",
        "md": "rounded",
        "lg": "rounded-lg",
        "full": "rounded-full"
    }

    const buttonStyles = {
        normal: `border ${radiusStyles[radius] ?? radiusStyles["full"]} bg-sky-400 pr-7 pl-7 pt-2 pb-2 hover:bg-sky-600 transition ease-in-out`,
        clicked: `border ${radiusStyles[radius] ?? radiusStyles["full"]} bg-sky-200 pr-7 pl-7 pt-2 pb-2`,
        disable: `border ${radiusStyles[radius] ?? radiusStyles["full"]} bg-sky-700 pr-5 pl-5 pt-2 pb-2`,
    };

    const buttonSizes = {
        "sm": buttonStyles.normal + " pr-3 pl-3 pt-1 pb-1 text-sm",
        "md": buttonStyles.normal,
        "lg": buttonStyles.normal + " pr-10 pl-10 pt-5 pb-5 text-lg",
    }

    return (
        <button disabled={isDisabled} className={
            isDisabled ? buttonStyles.disable: buttonSizes[size] ?? buttonSizes["md"]
        }
        onClick={(e) => {
            e.target.className = buttonStyles.clicked;
            setTimeout(() => {
                e.target.className = buttonSizes[size] ?? buttonSizes["md"];
            }, 50);
            this?.handleClick(e);
        }}
        {...defaultProps}
        >{text}</button>
    )
}

Button.propTypes = {
    
}
