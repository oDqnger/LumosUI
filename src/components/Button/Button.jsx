import PropTypes from 'prop-types'

export function Button(props) {

    const {
        color="default",
        radius="lg",
        onClick: handleClick,
        size="md", 
        isDisabled=false, 
        children: text, 
        className,
        ...defaultProps
        } = props;


    const buttonColors = {
        "default": {
            normal: "bg-zinc-500",
            clicked: "bg-zinc-200",
            disable: "bg-zinc-700",
            hover: "hover:bg-zinc-600"
        },
        "primary": {
            normal: "bg-sky-500",
            clicked: "bg-sky-200",
            disable: "bg-sky-700",
            hover: "hover:bg-sky-600"
        },
        "secondary": {
            normal: "bg-violet-600",
            clicked: "bg-violet-300",
            disable: "bg-violet-800",
            hover: "hover:bg-violet-700"
        },
        "success": {
            normal: "bg-green-500",
            clicked: "bg-green-200",
            disable: "bg-green-700",
            hover: "hover:bg-green-600"
        },
        "warning": {
            normal: "bg-amber-600",
            clicked: "bg-amber-300",
            disable: "bg-amber-800",
            hover: "hover:bg-amber-700"
        },
        "danger": {
            normal: "bg-red-600",
            clicked: "bg-red-300",
            disable: "bg-red-800",
            hover: "hover:bg-red-700"
        }
    }
    
    const radiusStyles = {
        "none": "rounded-none",
        "sm": "rounded-sm",
        "md": "rounded",
        "lg": "rounded-lg",
        "full": "rounded-full"
    }

    const buttonStyles = {
        normal: `border ${radiusStyles[radius] ?? radiusStyles["full"]} ${buttonColors[color].normal ?? buttonColors["default"].normal} pr-7 pl-7 pt-2 pb-2 ${buttonColors[color].hover ?? buttonColors["default"].hover} transition ease-in-out ${className ?? ""}`,
        clicked: `border ${radiusStyles[radius] ?? radiusStyles["full"]} ${buttonColors[color].clicked ?? buttonColors["default"].clicked} pr-7 pl-7 pt-2 pb-2 ${className ?? ""}`,
        disable: `border ${radiusStyles[radius] ?? radiusStyles["full"]} ${buttonColors[color].disable ?? buttonColors["default"].disable} pr-7 pl-7 pt-2 pb-2 ${className ?? ""}`,
    };

    const buttonSizes = {
        "sm": "pr-3 pl-3 pt-1 pb-1 text-sm",
        "md": "",
        "lg": "pr-10 pl-10 pt-5 pb-5 text-lg",
    }

    return (
        <button disabled={isDisabled} className={
            isDisabled ? buttonStyles.disable + (buttonSizes[size] ?? buttonSizes["md"]): buttonStyles.normal + (buttonSizes[size] ?? buttonSizes["md"])
        }
        onClick={(e) => {
            e.target.className = buttonStyles.clicked + (buttonSizes[size] ?? buttonSizes["md"]);
            setTimeout(() => {
                e.target.className = buttonStyles.normal + (buttonSizes[size] ?? buttonSizes["md"]);
            }, 50);
            if (handleClick) {
                handleClick(e);
            }
        }}
        {...defaultProps}
        >{text}</button>
    )
}

Button.propTypes = {
    color: PropTypes.string,
    size: PropTypes.string,
    radius: PropTypes.string,
    onClick: PropTypes.func,
    isDisabled: PropTypes.bool,
}
