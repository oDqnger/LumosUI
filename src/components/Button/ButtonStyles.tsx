export const buttonColors = {
    "default": {
        normal: "bg-zinc-500",
        clicked: "bg-zinc-200",
        disable: "bg-zinc-700",
        hover: "hover:bg-zinc-600"
    },
    "primary": {
        normal: "bg-sky-500",
        clicked: "bg-sky-200",
        disable: "bg-sky-700",
        hover: "hover:bg-sky-600"
    },
    "secondary": {
        normal: "bg-violet-600",
        clicked: "bg-violet-300",
        disable: "bg-violet-800",
        hover: "hover:bg-violet-700"
    },
    "success": {
        normal: "bg-green-500",
        clicked: "bg-green-200",
        disable: "bg-green-700",
        hover: "hover:bg-green-600"
    },
    "warning": {
        normal: "bg-amber-600",
        clicked: "bg-amber-300",
        disable: "bg-amber-800",
        hover: "hover:bg-amber-700"
    },
    "danger": {
        normal: "bg-red-600",
        clicked: "bg-red-300",
        disable: "bg-red-800",
        hover: "hover:bg-red-700"
    }
}

export const radiusStyles = {
    "none": "rounded-none",
    "sm": "rounded-sm",
    "md": "rounded",
    "lg": "rounded-lg",
    "full": "rounded-full"
}

export const buttonSizes = {
    "sm": "pr-3 pl-3 pt-1 pb-1 text-sm",
    "md": "",
    "lg": "pr-10 pl-10 pt-5 pb-5 text-lg",
}
