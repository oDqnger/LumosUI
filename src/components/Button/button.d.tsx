type ColorTypes = "default" | "primary" | "secondary" | "success" | "warning" | "danger"
type RadiusTypes = "none" | "sm" | "md" | "lg" | "full"
type SizeTypes = "sm" | "md" | "lg"

interface ButtonProps {
    isDisabled?: boolean,
    onClick?: Function,
}

interface DefaultProps {
    children?: any,
    className?: any,
}

interface StyledProps {
    color?: ColorTypes,
    radius?: RadiusTypes,
    size?: SizeTypes,
}

export type AllProps = ButtonProps & DefaultProps & StyledProps;
