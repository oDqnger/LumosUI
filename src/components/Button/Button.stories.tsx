import type { Meta, StoryObj } from "@storybook/react"
import { Button } from "./Button"
import "../../styles/output.css"

const meta: Meta<typeof Button> = {
    component: Button,
};

export default meta;

type Story = StoryObj<typeof Button>;

export const NormalButton: Story = {
    render: () => <Button>Button</Button>,
};

export const DisabledButton: Story = {
    render: () => <Button isDisabled>Button</Button>
}

export const PrimaryButton: Story = {
    render: () => <Button color="primary">Button</Button>
}

export const SecondaryButton: Story = {
    render: () => <Button color="secondary">Button</Button>
}

export const WarningButton: Story = {
    render: () => <Button color="warning">Button</Button>
}

export const DangerButton: Story = {
    render: () => <Button color="danger">Button</Button>
}

export const SuccessButton: Story = {
    render: () => <Button color="success">Button</Button>
}

export const DefaultButton: Story = {
    render: () => <Button color="default">Button</Button>
}

export const RadiusNoneButton: Story = {
    render: () => <Button radius="none">Button</Button>
}

export const RadiusSmallButton: Story = {
    render: () => <Button radius="sm">Button</Button>
}

export const RadiusMediumButton: Story = {
    render: () => <Button radius="md">Button</Button>
}

export const RadiusLargeButton: Story = {
    render: () => <Button radius="lg">Button</Button>
}

export const RadiusFullButton: Story = {
    render: () => <Button radius="full">Button</Button>
}

export const SizeSmallButton: Story = {
    render: () => <Button size="sm">Button</Button>
}

export const SizeMediumButton: Story = {
    render: () => <Button size="md">Button</Button>
}

export const SizeLargeButton: Story = {
    render: () => <Button size="lg">Button</Button>
}

export const CustomButton: Story = {
    render: () => <Button className="m-5 text-lg w-96">Button</Button>
}
