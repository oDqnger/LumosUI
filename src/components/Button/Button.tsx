import React, { forwardRef } from "react";
import { AllProps } from "./button.d";
import { buttonColors, buttonSizes, radiusStyles, baseStyles } from "./ButtonStyles"

function Button(props: AllProps, ref) {

    const {
    color="default",
    radius="lg",
    onClick: handleClick,
    size="md", 
    isDisabled=false, 
    children: text, 
    className="",
    customStyles=null,
    overrideDefaultOnClick,
    ...defaultProps
    } = props;

    const isDisabledOrNormal = isDisabled ? buttonColors[color].disable : buttonColors[color].normal + buttonColors[color].hover;

    const combinedStyles =
    className
    +baseStyles
    +radiusStyles[radius]
    +buttonSizes[size]
    +isDisabledOrNormal

    return (
        <button
        disabled={isDisabled}
        className={customStyles ?? combinedStyles}
        onClick={(e) => {
            if (customStyles == null || overrideDefaultOnClick == null) {
                const target = e.target as HTMLButtonElement;
                target.className = combinedStyles.replace(buttonColors[color].normal, buttonColors[color].clicked);
                setTimeout(() => {
                    target.className = combinedStyles
                }, 40);
            }

            if (overrideDefaultOnClick != null) {
                overrideDefaultOnClick();
            }
            
            if (handleClick) {
                handleClick(e);
            }
        }}
        ref={ref}
        {...defaultProps}
        >{text}</button>
    )
}

export default forwardRef(Button);
