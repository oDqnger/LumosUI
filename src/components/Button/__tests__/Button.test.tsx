import '@testing-library/jest-dom/extend-expect';
import React from "react"
import { fireEvent, render } from "@testing-library/react"
import { Button } from "../Button"
import "@testing-library/jest-dom"

describe("Button component", () => {

    it("should match snapshot", () => {
        const { container } = render(<Button>Hello</Button>)
        expect(container).toMatchSnapshot();
    })

    it("render button component", () => {
        const { container } = render(<Button>Click me!</Button>)
        expect(container).toBeInTheDocument()
    })

    it("fires the onClick function", () => {
        let something = false;
        const { getByText } = render(<Button onClick={() => {
            something = true;
        }}>Click me!</Button>)
    

        // eslint-disable-next-line testing-library/prefer-screen-queries
        fireEvent.click(getByText("Click me!"));
        expect(something).toBe(true);
    })
})
