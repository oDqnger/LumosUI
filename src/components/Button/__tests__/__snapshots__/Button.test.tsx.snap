// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should match snapshot 1`] = `
<div>
  <button
    class="border rounded-lg bg-zinc-500 pr-7 pl-7 pt-2 pb-2 hover:bg-zinc-600 transition ease-in-out "
  >
    Hello
  </button>
</div>
`;
